### PR TITLE
Mark InPort and OutPort as InternaApi

### LINKED ISSUE #24146: InPort and OutPort can move to impl, after all
The [documentation for `InPort` and `OutPort`](https://github.com/akka/akka/blob/38a937ab16c99d1a547d89a5eb5d2d9d7dc3a1e3/akka-stream/src/main/scala/akka/stream/Shape.scala#L38) mentions

```
 * ... This type logically belongs
 * into the impl package but must live here due to how `sealed` works.
```

However, it can remain in this source file, making `sealed` work, but still be in the `impl` package.

akka-stream compiles with this change in the 2.5.7 tree against Scala 2.12.4.

```scala
// in Shape.scala
package akka.stream

// other imports...
import impl.{InPort, OutPort}

package impl {
  // InPort, OutPort appear nested in here
}

// rest of file follows
```

If there is some other reason not to do this, the documentation for `InPort` and `OutPort` could be updated to no longer claim that `sealed`'s restrictions are a reason to keep them in the `stream` package.

### PR DESCRIPTION
**Pull request by richardimaoka**
Refs #24146

The GitHub diff is not nicely showing the actual change.
Maybe it's easier to directly look at [my commit](https://github.com/richardimaoka/akka/blob/49a3da296b152234902efd2ae5debb9f675d7882/akka-stream/src/main/scala/akka/stream/Shape.scala) and the [previous commit](https://github.com/richardimaoka/akka/blob/009214ae07708e8144a279e71d06c4a504907e31/akka-stream/src/main/scala/akka/stream/Shape.scala).

I only:
- [put `package impl`](https://github.com/richardimaoka/akka/blob/49a3da296b152234902efd2ae5debb9f675d7882/akka-stream/src/main/scala/akka/stream/Shape.scala#L12) and [enclosed it](https://github.com/richardimaoka/akka/blob/49a3da296b152234902efd2ae5debb9f675d7882/akka-stream/src/main/scala/akka/stream/Shape.scala#L36)
- [modified comments](https://github.com/richardimaoka/akka/blob/49a3da296b152234902efd2ae5debb9f675d7882/akka-stream/src/main/scala/akka/stream/Shape.scala#L13L17) (same for `OutPort`)
- Changed `extends InPort` to [`extends impl.InPort`](https://github.com/richardimaoka/akka/blob/49a3da296b152234902efd2ae5debb9f675d7882/akka-stream/src/main/scala/akka/stream/Shape.scala#L86) (same for `OutPort`)

### CODE REVIEW

**Comment by richardimaoka** [comment]
will look into Jenkins failure (mima issues)

File: akka-stream/src/main/scala/akka/stream/Shape.scala [hunk 1]
```diff
@@ -9,55 +9,55 @@ import scala.collection.JavaConverters._
...
-sealed abstract class InPort { self: Inlet[_] ⇒
-  final override def hashCode: Int = super.hashCode
-  final override def equals(that: Any): Boolean = this eq that.asInstanceOf[AnyRef]
-
-  /**
-   * INTERNAL API
-   */
-  @volatile private[stream] var id: Int = -1
-
+package impl {
```
Comment (hepin1989) [inline akka-stream/src/main/scala/akka/stream/Shape.scala]: why not a dedicated folder? for the diff?

Comment (richardimaoka) [reply akka-stream/src/main/scala/akka/stream/Shape.scala]: Due to how `sealed` works. Classes extending `InPort`/`OutPort` need to be in the same file.

**Comment by richardimaoka** [comment]
Now Jenkins fails with this error which I am not familiar with:

```
[error] 	at java.lang.Thread.run(Thread.java:745)
[error] (whitesourceCheckPolicies) Error communicating with service: Server error: java.lang.NullPointerException
```

Locally on my PC, it is currently failing with quite a lot of mima errors like below...if no way to get around this, we should close the original issue without fixing it?

```
[info] akka-stream: found 0 potential binary incompatibilities while checking against com.typesafe.akka:akka-stream_2.11:2.4.4  (filtered 538)
[info] akka-stream: found 0 potential binary incompatibilities while checking against com.typesafe.akka:akka-stream_2.11:2.4.3  (filtered 541)
[info] akka-stream: found 71 potential binary incompatibilities while checking against com.typesafe.akka:akka-stream_2.11:2.5.7  (filtered 78)
[error]  * method assignPort(akka.stream.InPort,Int,java.lang.Object)Unit in class akka.stream.impl.SinkModulePhase in current version does not have a correspondent with same parameter signature among (akka.stream.impl.InPort,Int,java.lang.Object)Unit, (akka.stream.impl.OutPort,Int,java.lang.Object)Unit
[error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.stream.impl.SinkModulePhase.assignPort")
[error]  * method assignPort(akka.stream.OutPort,Int,java.lang.Object)Unit in class akka.stream.impl.SinkModulePhase in current version does not have a correspondent with same parameter signature among (akka.stream.impl.InPort,Int,java.lang.Object)Unit, (akka.stream.impl.OutPort,Int,java.lang.Object)Unit
[error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.stream.impl.SinkModulePhase.assignPort")
...
...
...
...
...
[error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.stream.impl.LinearTraversalBuilder.isUnwired")
[error]  * method this(akka.stream.InPort,akka.stream.OutPort,Int,Int,akka.stream.impl.Traversal,akka.stream.impl.TraversalBuilder,akka.stream.Attributes,akka.stream.impl.Traversal,akka.stream.impl.IslandTag)Unit in class akka.stream.impl.LinearTraversalBuilder's type is different in current version, where it is (akka.stream.impl.InPort,akka.stream.impl.OutPort,Int,Int,akka.stream.impl.Traversal,akka.stream.impl.TraversalBuilder,akka.stream.Attributes,akka.stream.impl.Traversal,akka.stream.impl.IslandTag)Unit instead of (akka.stream.InPort,akka.stream.OutPort,Int,Int,akka.stream.impl.Traversal,akka.stream.impl.TraversalBuilder,akka.stream.Attributes,akka.stream.impl.Traversal,akka.stream.impl.IslandTag)Unit
[error]    filter with: ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.stream.impl.LinearTraversalBuilder.this")
[error] java.lang.RuntimeException: akka-stream: Binary compatibility check failed!
[error] 	at scala.sys.package$.error(package.scala:27)
[error] 	at com.typesafe.tools.mima.plugin.SbtMima$.reportModuleErrors(SbtMima.scala:114)
[error] 	at com.typesafe.tools.mima.plugin.MimaPlugin$.$anonfun$mimaReportSettings$11(MimaPlugin.scala:78)
[error] 	at com.typesafe.tools.mima.plugin.MimaPlugin$.$anonfun$mimaReportSettings$11$adapted(MimaPlugin.scala:61)
[error] 	at scala.collection.immutable.HashMap$HashMap1.foreach(HashMap.scala:231)
[error] 	at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:462)
[error] 	at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:462)
[error] 	at com.typesafe.tools.mima.plugin.MimaPlugin$.$anonfun$mimaReportSettings$9(MimaPlugin.scala:61)
[error] 	at com.typesafe.tools.mima.plugin.MimaPlugin$.$anonfun$mimaReportSettings$9$adapted(MimaPlugin.scala:45)
[error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:44)
[error] 	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:39)
[error] 	at sbt.std.Transform$$anon$4.work(System.scala:66)
[error] 	at sbt.Execute.$anonfun$submit$2(Execute.scala:262)
[error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:16)
[error] 	at sbt.Execute.work(Execute.scala:271)
[error] 	at sbt.Execute.$anonfun$submit$1(Execute.scala:262)
[error] 	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:174)
[error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:36)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
[error] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
[error] 	at java.lang.Thread.run(Thread.java:745)
[error] (mimaReportBinaryIssues) akka-stream: Binary compatibility check failed!
```

**Comment by raboof** [comment]
The whitesource error should be a transient one, see #24499

**Comment by richardimaoka** [comment]
I see, but I think the remaining Mima issues are real ... :worried:

File: akka-stream/src/main/scala/akka/stream/Shape.scala (hunk 1 above)
Comment (hepin1989) [reply akka-stream/src/main/scala/akka/stream/Shape.scala]: I mean split the `Inport` and `OutPort`

**Comment by richardimaoka** [comment]
Due to the mima issues, I think I should close this PR and we also close the original issue #24146 without a fix?

**Review by jrudolph — COMMENTED** [review]
I wonder if we actually care about the class hierarchy being sealed? If not we could just move those two classes into their own file in the impl package.

Most of the mima warnings are about internal usages, so we don't really care. However, there's also the warning that requires this filter:

```
ProblemFilters.exclude[MissingClassProblem]("akka.stream.InPort")
// same for OutPort
```

In the warning, it says "incompatible class hierarchy change". I wonder if that's really an issue or not if the moved superclass didn't have any publicly accessible methods.

**Comment by richardimaoka** [comment]
Sorry for taking long on this. I rebased the branch as it got outdated for long.

> If not we could just move those two classes into their own file in the impl package.

Done in the new commit.

> I wonder if that's really an issue or not if the moved superclass didn't have any publicly accessible methods.

As in the code diff, the super classes `InPort`/`OutPort` only have private methods (except `hashCode` and `equals`). 

```scala
abstract class InPort { self: Inlet[_] ⇒
  final override def hashCode: Int = super.hashCode
  final override def equals(that: Any): Boolean = this eq that.asInstanceOf[AnyRef]
  @volatile private[stream] var id: Int = -1
  @volatile private[stream] var mappedTo: InPort = this
  private[stream] def inlet: Inlet[_] = this
}
```

However, sub classes `Inlet` and `Outlet` are used in user-defined custom GraphStage, and they are changed as follows. Does that cause bin-compat errors for users? If we are not sure I can do quick experiment to see how it goes.

```scala
-final class Inlet[T] private (val s: String) extends InPort {
+final class Inlet[T] private (val s: String) extends impl.InPort {
```

**Comment by richardimaoka** [comment]
Test passed on another rebase.

**Comment by johanandren** [comment]
Actually, after looking at it I don't think we can safely do this, given that we didn't document `InPort` and `OutPort` as internal API it is possible (although unlikely) that someone coded against it and we'd break source compatibility through this change.

I think we should mark it as internal API and fix the next time we can break compatibility.

**Comment by richardimaoka** [comment]
OK reflected the review comments. Is there a way to mark the original issue as something we should tackle in the future or should it just be closed as is?

**Comment by johanandren** [comment]
Some times I think we mark issues with the next major version as milestone (3.0 in this case then), when we know we need to break bincomp

**Comment by johanandren** [comment]
I'll do that mark for the future with the ticket and close this PR then, thanks for the work @richardimaoka, sorry it ended up being too complicated for now.

**Pull request closed without merging**